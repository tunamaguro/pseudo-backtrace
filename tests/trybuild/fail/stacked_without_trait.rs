use pseudo_backtrace::StackError;

#[derive(Debug)]
struct NotStacked(std::io::Error);
impl core::fmt::Display for NotStacked {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        self.0.fmt(f)
    }
}
impl core::error::Error for NotStacked {}

#[derive(Debug, StackError)]
struct BadOuter {
    #[stack_error(stacked)]
    source: NotStacked,
    #[location]
    location: &'static core::panic::Location<'static>,
}

impl core::fmt::Display for BadOuter {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        "bad".fmt(f)
    }
}
impl core::error::Error for BadOuter {}

fn main() {}
