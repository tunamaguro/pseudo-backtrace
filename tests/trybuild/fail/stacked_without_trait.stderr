error[E0599]: the method `as_dyn_stack_error` exists for struct `NotStacked`, but its trait bounds were not satisfied
  --> tests/trybuild/fail/stacked_without_trait.rs:12:17
   |
 4 | struct NotStacked(std::io::Error);
   | ----------------- method `as_dyn_stack_error` not found for this struct because it doesn't satisfy `NotStacked: AsDynStackError<'_>` or `NotStacked: StackError`
...
12 | #[derive(Debug, StackError)]
   |                 ^^^^^^^^^^
   |
   = note: the following trait bounds were not satisfied:
           `NotStacked: StackError`
           which is required by `NotStacked: AsDynStackError<'_>`
note: the trait `StackError` must be implemented
  --> src/lib.rs
   |
   | pub trait StackError: core::error::Error {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `as_dyn_stack_error`, perhaps you need to implement it:
           candidate #1: `AsDynStackError`
   = note: this error originates in the derive macro `StackError` (in Nightly builds, run with -Z macro-backtrace for more info)
